ifgate/ifmail.8
.\" Set up \*(lq, \*(rq if -man hasn't already set it up.
.if @@\*(lq@ \{\
.	ds lq "
.	if t .ds lq ``
.	if !@@\(lq@ .ds lq "\(lq
.\}
.if @@\*(rq@ \{\
.	ds rq "
.	if t .ds rq ''
.	if !@@\(rq@ .ds rq "\(rq
.\}
.de Id
.ds Rv \\$3
.ds Dt \\$4
..
.de Sp
.if n .sp
.if t .sp 0.4
..
.TH IFMAIL 8 "\*(Dt" "Internet-Fidonet Gateway"
.UC 4
.SH NAME
.B	ifmail
\- Convert RFC-822 and RFC-1036 mail/news to FTN Fidonet packets.
.PP
.B	ifnews
\- Convert RFC-1036 Usenet news to FTN Fidonet packets.
.PP
.br
.SH Synopsis
.B	ifmail
[-h]  [-x verbosity]  [-I file]  [-n]  [-r Address] [-g grade]  receipent ...
.PP
.B	ifnews
[-h]  [-x verbosity]  [-I file]  [-r Address] [-g grade]  [receipent]
.PP
.SH Description
.B Ifmail
converts RFC-822 compliant mail and RFC-1036 compliant 
news to FTN packets. It must run setuid owner of fidonet related stuff. 
The current packet naming is BinkleyTerm, with file/directory names in 
lowcase. Naming conventions are implemented in a separate source file, 
pktname.c, to make modifications easier.
.PP
The default mode is the mail mode of operation. In mail mode, you must
specify recepient address(es) on the command line, in the form: 
.br
"Eugene.Crosser@p14.f6.n5020.z2.fidonet.org".
.PP
The first recepient address is used as the routing address, if no routing
address is specified.  In news mode then the $NEWSSITE environment 
variable is used as the routing address.
.PP
When the news mode is in effect, (i.e. "-n" key is specified or the 
program is called by alias "ifnews") a news article (or batch) is
expected on stdin. 
.PP
.B Ifmail 
will try to preserve as much information as possible 
in "X-FTN-..." headers and "^ARFC-..." kludges, and restore messages more 
or less accurately to their original form when double-gatewaying.
.PP
A dbm-based alias database is supported, so if a message passes from a
newsgroup to an echo, the author's free form name and domain address are 
stored, and when a netmail reply comes from fidonet to that free form name, 
it is passed as a mail message to the remembered domain address.
.PP
.B ATTENTION: 
This mechanism will not work if you specify fidonet-style
address as the "visible name" in your MTA.  Do not do it!  Even if you
don't have an official Internet domain, use something.uucp as your
primary machine name, NOT fN.nM.zL.fidonet.org.
.PP
Several addresses may be specified in the config file, netmail from fidonet 
addressed to any of those addresses is assumed local and resolved through 
the aforementioned database, otherwise the mail will be routed through the 
normal MTA (and presumably packed to some other fidonet node). If there is 
a "To:" line at the beginning of the fido message, the address is taken from
it 
.B (ONLY THE FIRST ADDRESS!).
.PP
When an RFC message is split, unique MSGIDs are generated for all parts 
after the first one. In any case, the original "Message-ID:" header is 
preserved in the "^ARFC-Message-ID:" kludge and used if the message comes 
back to usenet on some (other) gateway.
.PP
On the way from news to echo, if the node to which the packet is being
created is present in an "X-FTN-SEEN-BY:" header, the message is not
included in the packet. SEEN-BY lines in the messages included into
the packet consist of (1) your node primary address, (2) copy of
"X-FTN-SEEN-BY:" header(s), (3) ftn addresses that could be parsed
from the CNews file "$NEWSCTL/log" in the line with the corresponding
Message-ID. For the latter to work, you should have a Cnews compatible
log file available for reading and have an "ndbm" package. This works
with INN too.
.PP
.B ATTENTION:
your feed name in the cnews "sys" file should be fNNN.nMMM or
pNNN.fMMM.nLLL, without zone and domain, see the examples included
with the source distribution.
.PP
In some cases, though, you will need to specify zone and/or domain of
the feed, e.g. if you are exporting echomail to several networks.  In
such case, specify the "cutdown" fNNN.nMMM notation after the slash, to
prevent exporting back (Refer the your news systems manual).  This trick
may also be useful if your hub presents non-primary AKA in the echomail 
it gives you.
.PP
.B Ifmail 
does make some attempt to process file attaches, but only locally.
Not passing to the Internet and back, but if a fileattach netmail 
message is routed from one fidonet node to another, it probably will
take the attached file with it.
.PP
There is also a feature to define "forbidden" groups. If a usenet
message is crossposted to some of the gated groups _and_ to some
of the forbidden groups it will not be passed to any fidonet echoes.
Messages with "Control:" headers are also not passed to FidoNet.
.PP
.SH OPTIONS
.B \-h
	Display a short help message.
.PP
.B \-x verbosity
	Set the debug verbosity
.B verbosity
may be a number from 0 to 32 to set 'on' bits from 1 to number, or
a string of letters where 'a' = bit 1, 'b' = bit 2, etc. up to bit 26.
.PP
.B \-I File
	Use the alternate configuration file
.B File.
.PP
.B \-n
	Set news mode for processing RFC-1036 Usenet news rather than RFC-822
electronic mail.
.PP
.B \-r Address
	Route packets to the Fidonet address 
.B Address.
.B Address
should be in the format "[pNN.]fNN.nNN[.zNN[.domain]]".
.PP
.B \-g Grade
	Set the type of Fidonet packet to create.  Where 
.B Grade
is one of:
.br
	N = Normal (Default)
.br
	C = Crash
.br
	H = Hold
.PP

.SH FILES
.B $IFLIBDIR/config
	Runtime configuration file.
.PP
.SH SEE ALSO
ifcico(8), ifpack(8), ifunpack(8), iftoss(8)
.PP
.SH Acknowledgements
Some ideas taken from Fidogate/RFmail package, written by Teemu Torma
and hacked by Martin Junius.  Some modeules taken from INN package.
Thanks to Michael Bravo <mbravo@tctube.spb.su> (who was the first) and
many others for testing.
.PP
.if t \(co
1993, 1994 Eugene Crosser
.PP
This is free software. You can do what you wish with it
as long as this copyright notice is preserved.


ifgate/ifmail.c
#include <sys/types.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <sys/stat.h>
#include <sysexits.h>
#ifdef HAS_SYSLOG
#include <syslog.h>
#endif
#include "mygetopt.h"
#include "lutil.h"
#include "xutil.h"
#include "ftn.h"
#include "rfcaddr.h"
#include "falists.h"
#include "rfcmsg.h"
#include "ftnmsg.h"
#include "areas.h"
#include "config.h"
#include "version.h"
#include "trap.h"
#include "hash.h"

extern int newsmode;
extern int suppress_received;
extern char *configname;
extern char passwd[];

extern int putmessage(rfcmsg *,ftnmsg *,FILE *,faddr*,char,fa_list **);
extern void closepkt(void);
extern char *bgets(char *,int,FILE *);
extern void readareas(char *);
extern void try_attach(char *,int,faddr *,char);
extern int flag_on(char *,char *);
extern unsigned INT16 crc(char *);
extern void close_alias_db(void);
#ifdef HAS_NDBM_H
extern void close_id_db(void);
#endif

int usetmp=0;
faddr *bestaka;
extern int fakeoutbound;

void usage(void)
{
	confusage("-N -o<flavors> -n -r<addr> -g<grade> <recip> ...");
	fprintf(stderr,"-N		put packets to /tmp/ifmail directory\n");
	fprintf(stderr,"-o<flavors>	force `out' mode for these flavors\n");
	fprintf(stderr,"-o+		force `out' mode for all flavors\n");
	fprintf(stderr,"-o-		reset `out' mode for all flavors\n");
	fprintf(stderr,"-n		set news mode\n");
	fprintf(stderr,"-r<addr>	address to route packet\n");
	fprintf(stderr,"-g<grade>	[ n | c | h ] \"flavor\" of packet\n");
	fprintf(stderr,"<recip>		list of receipient addresses\n");
}

int main(argc,argv)
int argc;
char *argv[];
{
	int c;
	char *p;
	char buf[BUFSIZ];
	FILE *fp;
	char *routec=NULL;
	faddr *route = NULL;
	fa_list **envrecip, *envrecip_start = NULL;
	int envrecip_count=0;
	area_list *area = NULL, *area_start = NULL;
	int area_count=0;
	int msg_in=0,msg_out=0;
	rfcmsg *msg=NULL,*tmsg;
	ftnmsg *fmsg=NULL;
	faddr *taddr;
	char cflavor='\0',flavor;
	fa_list *sbl = NULL;
	unsigned long svmsgid;
	char *outmode=NULL;

#if defined(HAS_SYSLOG) && defined(MAILLOG)
	logfacility=MAILLOG;
#endif

	setmyname(argv[0]);
	catch(myname);
       while ((c=getopt(argc,argv,"g:r:x:I:RnNho")) != -1)
	if (confopt(c,optarg)) switch (c)
	{
		case 'N':	fakeoutbound=1; break;
		case 'o':	outmode=optarg; break;
		case 'g':	if (optarg && ((*optarg == 'n') || 
				   (*optarg == 'c') || (*optarg == 'h')))
				{
					cflavor=*optarg;
				}
				else 
				{
					usage(); 
					exit(EX_USAGE);
				}
				break;
		case 'r':	routec=optarg; break;
		case 'n':	newsmode=1; break;
               case 'R':       suppress_received=1; break;
		default:	usage(); exit(EX_USAGE);
	}

	if (cflavor == 'n') cflavor='o';

	if (readconfig())
	{
		fprintf(stderr,"Error getting configuration, aborting\n");
		exit(EX_DATAERR);
	}

	if (outmode)
	{
		if (*outmode == '-')
			nonpacked="";
		else if (*outmode == '+')
			nonpacked="och";
		else
			nonpacked=outmode;
	}
	if (nonpacked == NULL) nonpacked="";
	for (p=nonpacked;*p;p++)
	if ((*p == 'f') || (*p == 'n')) *p='o';
	/* constant values will not be modified anyway */

	if ((routec) && ((route=parsefaddr(routec)) == NULL))
		logerr("unparsable route address \"%s\" (%s)",
			S(routec),addrerrstr(addrerror));

	if ((p=strrchr(argv[0],'/'))) p++;
	else p=argv[0];
	if (!strcmp(p,"ifnews")) newsmode=1;

	if (newsmode)
	{
		readareas(areafile);
#if defined(HAS_SYSLOG) && defined(NEWSLOG)
		logfacility=NEWSLOG;
#endif
	}
	else
	{
		if (strchr(nonpacked,'m'))
			nonpacked="och";
	}

	envrecip=&envrecip_start;
	while (argv[optind])
	if ((taddr=parsefaddr(argv[optind++]))) 
	{
		(*envrecip)=(fa_list*)xmalloc(sizeof(fa_list));
		(*envrecip)->next=NULL;
		(*envrecip)->addr=taddr;
		envrecip=&((*envrecip)->next);
		envrecip_count++;
	}
	else logerr("unparsable recipient \"%s\" (%s), ignored",
		argv[optind-1],
		addrerrstr(addrerror));

	if ((!newsmode) && (!envrecip_count))
	{
		logerr("No valid receipients specified, aborting");
		exit(EX_NOUSER);
	}

	if (!route && newsmode && envrecip_count)
		route=envrecip_start->addr;

	if (!route)
		if ((routec=getenv("NEWSSITE")))
			route=parsefaddr(routec);

	if (!route)
	{
		logerr("Routing address not specified, aborting");
		exit(EX_USAGE);
	}

	if (fakeoutbound)
	{
		mkdir("/tmp/ifmail",0777);
		loginf("packets will go to /tmp/ifmail");
	}

	bestaka=bestaka_s(route);

	for(envrecip=&envrecip_start;*envrecip;envrecip=&((*envrecip)->next))
		loginf("envrecip: %s",ascfnode((*envrecip)->addr,0x7f));
	loginf("route: %s",ascfnode(route,0x1f));

	umask(066); /* packets may contain confidential information */

	while (!feof(stdin))
	{
		usetmp=0;
		tidyrfc(msg);
		msg=parsrfc(stdin);

		flavor=cflavor;

		if (newsmode)
		{
			if (!flavor) flavor='o';
			tidy_arealist(area_start);
			area_start=areas(hdr("Newsgroups",msg));
			area_count=0;
			for(area=area_start;area;area=area->next)
			{
				area_count++;
				debug(9,"area: %s",S(area->name));
			}
			tidy_falist(&sbl);
			for (tmsg=msg;tmsg;tmsg=tmsg->next)
				if (strcasecmp(tmsg->key,"X-FTN-SEEN-BY") == 0)
					fill_list(&sbl,tmsg->val,NULL);
		}
		else
		{
			if ((p=hdr("X-FTN-FLAGS",msg)))
			{
				if (!flavor)
				{
					if (flag_on("CRS",p)) flavor='c';
					else if (flag_on("HLD",p)) flavor='h';
					else flavor='o';
				}
				if (flag_on("ATT",p))
					try_attach(hdr("Subject",msg),
						0,route,flavor);
				if (flag_on("TFS",p))
					try_attach(hdr("Subject",msg),
						1,route,flavor);
				if (flag_on("KFS",p))
					try_attach(hdr("Subject",msg),
						2,route,flavor);
			}
			if ((!flavor) && ((p=hdr("Priority",msg)) ||
			         (p=hdr("Precedence",msg)) ||
			         (p=hdr("X-Class",msg))))
			{
				while (isspace(*p)) p++;
				if ((strncasecmp(p,"fast",4) == 0) ||
				    (strncasecmp(p,"high",4) == 0) ||
				    (strncasecmp(p,"crash",5) == 0) ||
				    (strncasecmp(p,"airmail",5) == 0) ||
				    (strncasecmp(p,"special-delivery",5) == 0) ||
				    (strncasecmp(p,"first-class",5) == 0))
					flavor='c';
				else if ((strncasecmp(p,"slow",4) == 0) ||
				         (strncasecmp(p,"low",3) == 0) ||
				         (strncasecmp(p,"hold",4) == 0) ||
				         (strncasecmp(p,"news",4) == 0) ||
				         (strncasecmp(p,"bulk",4) == 0) ||
				         (strncasecmp(p,"junk",4) == 0))
					flavor='h';
			}
			if (!flavor) flavor='o';
		}

		if (((!newsmode) && (envrecip_count > 1)) ||
		    ((newsmode) && (area_count > 1)))
		{
			if ((fp=tmpfile()) == NULL)
			{
				logerr("$Cannot open temporary file");
				exit(EX_OSERR);
			}
			while(bgets(buf,sizeof(buf)-1,stdin))
				fputs(buf,fp);
			rewind(fp);
			usetmp=1;
		}
		else
		{
			fp=stdin;
			usetmp=0;
		}

		if (newsmode && ((area_count < 1) ||
		    (maxgroups && (group_count > maxgroups)) ||
		    hdr("Control",msg) ||
		     (in_list(route,&sbl))))
		{
			debug(9,"skipping news message");
			while(bgets(buf,sizeof(buf)-1,fp));
		}
		else
		{
			tidy_ftnmsg(fmsg);
			if ((fmsg=mkftnhdr(msg)) == NULL)
			{
				logerr("Unable to create FTN headers from RFC ones, aborting");
				exit(EX_UNAVAILABLE);
			}

			if (newsmode)
			{
				fill_list(&sbl,ascfnode(bestaka,0x06),NULL);
				fill_list(&sbl,ascfnode(route,0x06),NULL);
				fmsg->to->zone=route->zone;
				fmsg->to->net=route->net;
				fmsg->to->node=route->node;
				fmsg->to->point=route->point;
			}

			if (!newsmode)
			for(envrecip=&envrecip_start;*envrecip;envrecip=&((*envrecip)->next))
			{
				fmsg->to=(*envrecip)->addr;
				if (putmessage(msg,fmsg,fp,route,flavor,&sbl))
				{
					logerr("Unable to put netmail message into the packet, aborting");
					exit(EX_CANTCREAT);
				}
				if (usetmp) rewind(fp);
				fmsg->to=NULL;
				msg_out++;
			}
			else
			for(area=area_start;area;area=area->next)
			{
				fmsg->area=area->name;
				svmsgid=fmsg->msgid_n;
				hash_update_s(&fmsg->msgid_n,
						area->name);
				if (putmessage(msg,fmsg,fp,route,flavor,&sbl))
				{
					logerr("Unable to put echo message into the packet, aborting");
					exit(EX_SOFTWARE);
				}
				if (usetmp) rewind(fp);
				fmsg->area=NULL;
				fmsg->msgid_n=svmsgid;
				msg_out++;
			}
			msg_in++;
		}
		if (usetmp) fclose(fp);
	}

	closepkt();
	close_alias_db();
#ifdef HAS_NDBM_H
	close_id_db();
#endif

	loginf("end input %d, output %d messages",msg_in,msg_out);

	return EX_OK;
}
